opt: alias `@primary` to PRIMARY KEY

otan opened the pull request:
Release note (sql change): PRIMARY KEYs have been renamed to conform
to postgres (e.g. `@tbl_col1_col2_pkey`) in this release. To protect
certain use cases of backwards compatability, we also allow `@primary`
index hints to alias to the PRIMARY KEY, but only if no other index is
named `primary`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72534)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with some nits. Thanks!

[nit] "compatibility" in the release note

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72534#-:-Mo0IzLQ9Kxdk_2rQtAU:b-cpfwoz)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @RaduBerinde)

___
*[pkg/sql/logictest/testdata/logic_test/table, line 862 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72534#-Mo0JgBR4UWjryoYDoyA:-Mo0JgBR4UWjryoYDoyB:b2iafxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/48697b52fede7dbc20fb669c2060fd8d63c6a72d/pkg/sql/logictest/testdata/logic_test/table#L862)):*
> ```
>   id INT PRIMARY KEY,
>   b INT
> ); CREATE TABLE tbl_with_primary_named_index (
> ```

[nit] this is hard to read, just use two `statement ok`
___
*[pkg/sql/logictest/testdata/logic_test/table, line 883 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72534#-Mo0JVmEBcdmKOXdVkS_:-Mo0JVmEBcdmKOXdVkSa:b-n5ifz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/48697b52fede7dbc20fb669c2060fd8d63c6a72d/pkg/sql/logictest/testdata/logic_test/table#L883)):*
> ```
> 
> query T
> EXPLAIN SELECT b FROM tbl_with_primary_named_index@primary
> ```

[nit] we try to keep EXPLAIN out of logictests, maybe this should be an execbuilder test?
___
*[pkg/sql/logictest/testdata/logic_test/table, line 890 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72534#-Mo0Jx1K0oPLOByDCFm_:-Mo0Jx1K0oPLOByDCFma:bl0nchd) ([raw file](https://github.com/cockroachdb/cockroach/blob/48697b52fede7dbc20fb669c2060fd8d63c6a72d/pkg/sql/logictest/testdata/logic_test/table#L890)):*
> ```
> • scan
>   missing stats
>   table: tbl_with_primary_named_index@primary
> ```

This is not conclusive evidence that we're using the secondary index. Maybe add `ORDER BY b` or `SELECT * ` (to force an index join)
___
*[pkg/sql/opt/cat/utils.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72534#-Mo0JC2g1EDt08F1A95j:-Mo0JC2g1EDt08F1A95k:b-gu0kz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/48697b52fede7dbc20fb669c2060fd8d63c6a72d/pkg/sql/opt/cat/utils.go#L78)):*
> ```Go
> 		}
> 		// Fallback to referencing @primary as the PRIMARY KEY.
> 		// Note indexes with "primary" as their name takes precedence above.
> ```

[nit] note that or "Note: "


<!-- Sent from Reviewable.io -->

otan commented:
bors r=raduberinde

otan committed:
opt: alias `@primary` to PRIMARY KEY

Release note (sql change): PRIMARY KEYs have been renamed to conform
to postgres (e.g. `@tbl_col1_col2_pkey`) in this release. To protect
certain use cases of backwards compatibility, we also allow `@primary`
index hints to alias to the PRIMARY KEY, but only if no other index is
named `primary`.

otan commented:
bors r=raduberinde
